### DISCUSSION TITLE
UDP services w/traefik

### DISCUSSION THREAD

**Discussion by yaleman** [k3s-io/k3s#6655] [General]
I'm trying to expose a DNS server for testing, and I can't work out how to get the UDP traffic to the pod. I can successfully configure a TCP Ingress and can confirm it all works fine (port 15353). The UDP traffic never makes it to the pod - I've tried spinning up the exact same config with `tcpdump` running in a container and can prove the TCP SYN hits, but the UDP doesn't.

I've:
- configured the traefik helm chart to
  - add new entrypoints
  - add/expose the ports 
- configured services to point to the deployment
- configured ingresses to point to the services

.. what can I provide to help troubleshoot this? The nodes are running `v1.23.6`

**Comment by brandond** [comment]
Did you add a new DNS port to the chart config, or just add the listener to the additional args? You need to add the port and set it to exposed.

If you're reusing a port that's already exposed via TCP (and now want to expose it via both TCP and UDP), you might look at https://github.com/k3s-io/k3s/issues/5184#issuecomment-1292606245

**Reply by yaleman** [reply]
I did both
```
resource "kubernetes_manifest" "traefik_entrypoints" {
  manifest = {
    "apiVersion" = "helm.cattle.io/v1"
    "kind"       = "HelmChartConfig"
    "metadata" = {
      "name"      = "traefik"
      "namespace" = "kube-system"
    }
    spec = {
      "valuesContent" : <<-EOT
  additionalArguments:
  - "--entryPoints.goatnsudp.address=:15353/udp"
  - "--entryPoints.goatnstcp.address=:15353/tcp"
  ports:
    goatnsudp:
      port: 15353
      expose: true
      exposedPort: 15353
      protocol: UDP
    goatnstcp:
      port: 15353
      expose: true
      exposedPort: 15353
      protocol: TCP
EOT
    }
  }
}
```

**Reply by brandond** [reply]
Do you see one or both of those ports listed on the traefik service? What version of K3s (and what version of the Traefik chart) are you using?

**Reply by yaleman** [reply]
The traefik helm chart looks to be `10.19.300`

K3s `v1.23.6` per the original report

services look to be there (filtered for traefik):
```
NAMESPACE                  NAME                            TYPE           CLUSTER-IP      EXTERNAL-IP           PORT(S)                                      AGE
kube-system                traefik                         LoadBalancer   10.45.241.118   10.0.0.91,10.0.0.94   15353:30427/TCP,80:30824/TCP,443:32720/TCP   17d
kube-system                traefik-udp                     LoadBalancer   10.45.52.140    10.0.0.91,10.0.0.94   15353:31180/UDP                              17d
```

**Reply by yaleman** [reply]
Turns out the traefik Daemonset had set itself up with the tcp port, but not the UDP port. I edited it manually and it works now, but sounds like a bug that needs fixing.

**Reply by brandond** [reply]
I believe there is an issue with the traefik chart, where it will both add the UDP ports to the main service, and create the UDP service. Obviously only one can use them at a time, at least with ServiceLB. Does that match what you were seeing?

**Reply by yaleman** [reply]
It created the service, but didn't expose the port on the `traefik-6cb5865f8d-kplx9` pod.

That pod had the following labels:
```
app.kubernetes.io/instance=traefik
app.kubernetes.io/managed-by=Helm
app.kubernetes.io/name=traefik
helm.sh/chart=traefik-10.19.300
pod-template-hash=6cb5865f8d
```

**Reply by brandond** [reply]
It's not necessary for the port to be listed in the pod spec. All ports on the pod are exposed; the port list is mostly just for administrators and other discovery tools to consume. Services can target ports on pods by number regardless of whether or not they are listed. They only need to be listed to refer to them by name in other places.

**Reply by yaleman** [reply]
When I edited the deployment spec and added the port it instantly started working, so... maybe we're talking about different things?

**Reply by brandond** [reply]
its possible that the service is referring to them by name? I'm not super familiar with the details of how the chart sets things up. You might take a look at https://github.com/traefik/traefik-helm-chart and see if that issue has been reported and/or fixed yet.
